crates/fluvio-protocol/src/core/bytebuf.rs

```diff
@@ -0,0 +1,300 @@
+use std::io::Error;
+use std::io::ErrorKind;
+use std::ops::Deref;
+
+use bytes::buf::UninitSlice;
+use bytes::{Buf, Bytes, BufMut};
+
+use crate::{Encoder, Decoder, Version};
+
+/// Represnts a SmartModule WASM File bytes.
+///
+/// Provides a `Encoder` and `Decoder` implementation optimized for WASM files
+/// used in SmartModules.
+#[derive(Clone, Debug, Default, PartialEq, Eq)]
+pub struct ByteBuf {
+    inner: Vec<u8>,
```

Comment:
Since we may need to re-implement much of Bytes, we might just use https://docs.rs/bytes/1.2.1/bytes/struct.Bytes.html instead of `Vec<u8>`.  This also benefits cheap cloning instead of allocating a separate buffer.   This hopefully get rid of the unsafe implementation.

Reply:
As I can see in the docs for `BufMut` and `Bytes`, there's no implementation of `BufMut` for `Bytes`, we would have to implement it if we want to use `Bytes` as our inner type. https://docs.rs/bytes/latest/bytes/trait.BufMut.html#foreign-impls.

I have some work on `encoding`, `decoding` implementation using `Bytes` as our inner type, but as I can see we would still have to implement `BufMut` ourselves instead of taking advantage of the implementation provided by crate authors that we have with `Vec<u8>`.

Reply:
What about `BytesMut`?